### PR TITLE
Throw SparkDateTimeException for invalid cast in Spark3.3+ versions

### PR DESCRIPTION
**Pull request by nartal1**
This closes https://github.com/NVIDIA/spark-rapids/issues/6951.

I missed it by oversight in this PR https://github.com/NVIDIA/spark-rapids/pull/6938 . Since we didn't have UT which covers this case it was not flagged in pre-merge builds.  Verfied locally and the tests pass now.

### CODE REVIEW

**Comment by nartal1** [comment]
build

File: sql-plugin/src/main/330until340/scala/org/apache/spark/rapids/ShimTrampolineUtil.scala [hunk 1]
```diff
@@ -24,6 +24,6 @@ object ShimTrampolineUtil {
     // These are the arguments required by SparkDateTimeException class to create error message.
     val errorClass = "CAST_INVALID_INPUT"
     val messageParameters = Array("DOUBLE", "TIMESTAMP", SQLConf.ANSI_ENABLED.key)
-    new SparkDateTimeException(errorClass, Array(infOrNan) ++ messageParameters)
+    throw new SparkDateTimeException(errorClass, Array(infOrNan) ++ messageParameters)
```
Comment (jlowe) [inline sql-plugin/src/main/330until340/scala/org/apache/spark/rapids/ShimTrampolineUtil.scala:27]: Nit: If this is going to throw, it'd be nice if the method stated as such.  For example, RapidsErrorUtils creates the exception classes and returns them, it does not throw them, so there's precedent for methods to not throw the exceptions they are creating.

Alternatively, we could update the calling code to do the throw, which is what this originally seems intended to do.

Comment (nartal1) [reply sql-plugin/src/main/330until340/scala/org/apache/spark/rapids/ShimTrampolineUtil.scala]: Thanks @jlowe  for the suggestion. I have updated the calling code to do the throw. PTAL.

File: sql-plugin/src/main/330until340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala [hunk 2]
```diff
@@ -61,7 +62,7 @@ object RapidsErrorUtils extends RapidsErrorUtilsFor330plus {
     )
   }
 
-  def throwSparkDateTimeException(infOrNan: String): Unit = {
+  def throwSparkDateTimeException(infOrNan: String): SparkDateTimeException = {
```
Comment (jlowe) [inline sql-plugin/src/main/330until340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:65]: The method name implies it throws an exception but it doesn't.  Probably should be `sparkDateTimeException`.

File: sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala [hunk 3]
```diff
@@ -66,7 +67,7 @@ object RapidsErrorUtils extends RapidsErrorUtilsFor330plus {
     )
   }
 
-  def throwSparkDateTimeException(infOrNan: String): Unit = {
+  def throwSparkDateTimeException(infOrNan: String): SparkDateTimeException = {
```
Comment (jlowe) [inline sql-plugin/src/main/340+/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:70]: Same name-does-not-match-semantics issue here.

File: sql-plugin/src/main/330until340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala (hunk 2 above)
Comment (nartal1) [reply sql-plugin/src/main/330until340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala]: updated in [2421439](https://github.com/NVIDIA/spark-rapids/pull/6956/commits/2421439eb5cb9e6bfefc3d2280f0d273f2042638)

File: sql-plugin/src/main/330until340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala [hunk 4]
```diff
@@ -61,7 +62,7 @@ object RapidsErrorUtils extends RapidsErrorUtilsFor330plus {
     )
   }
 
-  def throwSparkDateTimeException(infOrNan: String): Unit = {
+  def sparkDateTimeException(infOrNan: String): SparkDateTimeException = {
     ShimTrampolineUtil.dateTimeException(infOrNan)
```
Comment (jlowe) [inline sql-plugin/src/main/330until340/scala/org/apache/spark/sql/rapids/shims/RapidsErrorUtils.scala:66]: Is it necessary to leverage `ShimTrampolineUtil` here?  The code here is under org.apache.spark.sql which is a subset of org.apache.spark, so not sure why we need to use org.apache.spark.rapids.himTrampolineUtil here.  Code would be simpler to read and smaller if we could avoid it.

**Comment by nartal1** [comment]
build

**Review by jlowe — APPROVED** [review]
Thanks for the extra polishing effort on this, @nartal1!

**Pull request merged**